File: 43173/backend/internal/model/revenue.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type RevenueType string

const (
	RevenueTypePlay      RevenueType = "play"
	RevenueTypeSubscription RevenueType = "subscription"
	RevenueTypeTicket    RevenueType = "ticket"
)

type RevenueRecord struct {
	ID          uint           `json:"id" gorm:"primaryKey"`
	UserID      uint           `json:"user_id" gorm:"index;not null"`
	ArtistID    uint           `json:"artist_id" gorm:"index;not null"`
	WorkID      *uint          `json:"work_id" gorm:"index"`
	OrderID     *uint          `json:"order_id" gorm:"index"`
	Type        RevenueType    `json:"type" gorm:"size:50;not null"`
	Amount      float64        `json:"amount"`
	PlayCount   int64          `json:"play_count"`
	Rate        float64        `json:"rate"`
	Status      int            `json:"status" gorm:"default:0"`
	Period      string         `json:"period" gorm:"size:20"`
	SettledAt   *time.Time     `json:"settled_at"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`
}

func (RevenueRecord) TableName() string {
	return "revenue_records"
}

type WithdrawStatus int

const (
	WithdrawStatusPending  WithdrawStatus = 0
	WithdrawStatusApproved WithdrawStatus = 1
	WithdrawStatusRejected WithdrawStatus = 2
	WithdrawStatusPaid     WithdrawStatus = 3
	WithdrawStatusFailed   WithdrawStatus = 4
)

type WithdrawRequest struct {
	ID            uint           `json:"id" gorm:"primaryKey"`
	UserID        uint           `json:"user_id" gorm:"index;not null"`
	ArtistID      uint           `json:"artist_id" gorm:"index;not null"`
	Amount        float64        `json:"amount"`
	Fee           float64        `json:"fee"`
	ActualAmount  float64        `json:"actual_amount"`
	Method        string         `json:"method" gorm:"size:50"`
	Account       string         `json:"account" gorm:"size:100"`
	AccountName   string         `json:"account_name" gorm:"size:100"`
	BankName      string         `json:"bank_name" gorm:"size:100"`
	Status        WithdrawStatus `json:"status" gorm:"default:0"`
	Remark        string         `json:"remark" gorm:"size:255"`
	ApprovedAt    *time.Time     `json:"approved_at"`
	ApprovedBy    *uint          `json:"approved_by"`
	RejectReason  string         `json:"reject_reason" gorm:"size:255"`
	PaidAt        *time.Time     `json:"paid_at"`
	TransactionNo string         `json:"transaction_no" gorm:"size:100"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `json:"-" gorm:"index"`
}

func (WithdrawRequest) TableName() string {
	return "withdraw_requests"
}

type Subscription struct {
	ID          uint           `json:"id" gorm:"primaryKey"`
	UserID      uint           `json:"user_id" gorm:"index;not null"`
	ArtistID    uint           `json:"artist_id" gorm:"index;not null"`
	Plan        string         `json:"plan" gorm:"size:50"`
	Amount      float64        `json:"amount"`
	StartDate   time.Time      `json:"start_date"`
	EndDate     time.Time      `json:"end_date"`
	Status      int            `json:"status" gorm:"default:1"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`
}

func (Subscription) TableName() string {
	return "subscriptions"
}

type DailyStats struct {
	ID             uint           `json:"id" gorm:"primaryKey"`
	Date           time.Time      `json:"date" gorm:"index"`
	UserID         uint           `json:"user_id" gorm:"index;not null"`
	ArtistID       uint           `json:"artist_id" gorm:"index;not null"`
	NewFollowers   int64          `json:"new_followers" gorm:"default:0"`
	NewPlays       int64          `json:"new_plays" gorm:"default:0"`
	NewLikes       int64          `json:"new_likes" gorm:"default:0"`
	NewShares      int64          `json:"new_shares" gorm:"default:0"`
	NewComments    int64          `json:"new_comments" gorm:"default:0"`
	Revenue        float64        `json:"revenue" gorm:"default:0"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `json:"-" gorm:"index"`
}

func (DailyStats) TableName() string {
	return "daily_stats"
}

type OperationLog struct {
	ID          uint           `json:"id" gorm:"primaryKey"`
	UserID      uint           `json:"user_id" gorm:"index"`
	Username    string         `json:"username" gorm:"size:50"`
	Module      string         `json:"module" gorm:"size:50"`
	Operation   string         `json:"operation" gorm:"size:100"`
	Method      string         `json:"method" gorm:"size:20"`
	Path        string         `json:"path" gorm:"size:255"`
	IP          string         `json:"ip" gorm:"size:45"`
	UserAgent   string         `json:"user_agent" gorm:"size:500"`
	Params      string         `json:"params" gorm:"type:text"`
	Result      string         `json:"result" gorm:"type:text"`
	Status      int            `json:"status"`
	Duration    int64          `json:"duration"`
	ErrorMsg    string         `json:"error_msg" gorm:"type:text"`
	CreatedAt   time.Time      `json:"created_at"`
	DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`
}

func (OperationLog) TableName() string {
	return "operation_logs"
}

File: 43173/frontend/src/api/revenue.ts
import request, { PageData, PageParams } from '@/utils/request'

export interface RevenueRecord {
  id: number
  user_id: number
  artist_id: number
  work_id?: number
  order_id?: number
  type: string
  amount: number
  play_count: number
  rate: number
  status: number
  period?: string
  settled_at?: string
  created_at: string
  updated_at: string
}

export interface WithdrawRequest {
  id: number
  user_id: number
  artist_id: number
  amount: number
  fee: number
  actual_amount: number
  method: string
  account: string
  account_name?: string
  bank_name?: string
  status: number
  remark?: string
  approved_at?: string
  approved_by?: number
  reject_reason?: string
  paid_at?: string
  transaction_no?: string
  created_at: string
  updated_at: string
}

export interface DailyStats {
  id: number
  date: string
  user_id: number
  artist_id: number
  new_followers: number
  new_plays: number
  new_likes: number
  new_shares: number
  new_comments: number
  revenue: number
  created_at: string
  updated_at: string
}

export interface OperationLog {
  id: number
  user_id?: number
  username?: string
  module: string
  operation: string
  method: string
  path: string
  ip: string
  user_agent: string
  params?: string
  result?: string
  status: number
  duration: number
  error_msg?: string
  created_at: string
}

export interface WithdrawParams {
  amount: number
  method: string
  account: string
  account_name?: string
  bank_name?: string
}

export const revenueApi = {
  getRecords: (params: PageParams & { start_date?: string; end_date?: string }) => 
    request.get<PageData<RevenueRecord>>('/revenue/records', params),
  getArtistRecords: (artistId: number, params: PageParams & { start_date?: string; end_date?: string }) => 
    request.get<PageData<RevenueRecord>>(`/revenue/records/artist/${artistId}`, params),
  getTotal: (params?: { start_date?: string; end_date?: string }) => 
    request.get<{ total_revenue: number }>('/revenue/total', params),
  getSummary: () => request.get('/revenue/summary'),
  
  requestWithdraw: (data: WithdrawParams) => request.post<WithdrawRequest>('/withdraw', data),
  getWithdrawList: (params: PageParams & { status?: number }) => 
    request.get<PageData<WithdrawRequest>>('/withdraw', params),
  getWithdrawStatusList: () => request.get('/withdraw/status-list'),
  
  getSubscriptions: (params: PageParams) => request.get('/subscriptions', params),
  getArtistSubscribers: (artistId: number, params: PageParams) => 
    request.get(`/subscriptions/artist/${artistId}`, params),
  
  getDailyStats: (params: { start_date: string; end_date: string }) => 
    request.get<DailyStats[]>('/stats/daily', params),
  getArtistDailyStats: (artistId: number, params: { start_date: string; end_date: string }) => 
    request.get<DailyStats[]>(`/stats/daily/artist/${artistId}`, params),
  getArtistStats: (artistId: number) => request.get(`/stats/artist/${artistId}`),
  
  getAdminWithdrawList: (params: PageParams & { status?: number }) => 
    request.get<PageData<WithdrawRequest>>('/admin/withdraw', params),
  approveWithdraw: (id: number) => request.put(`/admin/withdraw/${id}/approve`),
  rejectWithdraw: (id: number, data: { reason: string }) => 
    request.put(`/admin/withdraw/${id}/reject`, data),
  markWithdrawPaid: (id: number, data?: { transaction_no?: string }) => 
    request.put(`/admin/withdraw/${id}/paid`, data || {}),
  settleRevenue: (period: string) => request.post('/admin/revenue/settle', { period }),
  
  getOperationLogs: (params: PageParams & { user_id?: number; module?: string; keyword?: string }) => 
    request.get<PageData<OperationLog>>('/admin/operation-logs', params),
  
  exportRevenue: (params?: { start_date?: string; end_date?: string }) => 
    request.get('/export/revenue', params, { responseType: 'blob' }),
  exportWithdraw: (params?: { status?: number }) => 
    request.get('/export/withdraw', params, { responseType: 'blob' })
}
